データベース設計

sho1999 opened the pull request:
# what
データベース設計をREADMEに書きました

# why
開発をスムーズにするため

sho1999 committed:
ER図作成

sho1999 committed:
データベース設計

yuri240 commented:
レビュー依頼ありがとうございます！

ご提出いただきましたデータベース設計のプルリクエストですが、
こちらは**ER図**とご一緒にレビューさせていただいております。

つきましては、
**ER図に記載されている内容が確認できるキャプチャ画像**
を添付の上、再度ご提出いただけますでしょうか？

「Gyazo　使い方」とgoogleで検索し、
ツールによって生成されたURLをこちらの
プルリクエスト欄に貼っていただければ幸いです。

お手数ではありますが、何卒よろしくお願い致します。

sho1999 commented:
![スクリーンショット 2021-04-10 14 50 02](https://user-images.githubusercontent.com/63441286/114259849-0939cc80-9a0c-11eb-9bbc-f862bdadd6ab.jpg)
こちらERず図になります

yuri240 commented on README.md:
```diff
@@ -1,24 +1,70 @@
+- belongs_to :user
+
+
+## purchaseテーブル
+
+| Column        | Type       | Options                        |
+| ------------- | ---------- | ------------------------------ |
+| card_info     | string     | null: false                    |
+| term          | string     | null: false                    |
+| security_code | string     | null: false                    |
+| user_id       | references | null: false, foreign_key: true |
+
+
+### Association
+
+- belongs_to :user
+
```
purchaseテーブルに用意していただきたいカラムは以下となります。
```
purchaseテーブル：誰が出品した商品を購入したかを管理するテーブル
必要なカラム↓
①userカラム(外部キー)
②itemカラム(外部キー)
```

yuri240 commented on README.md:
```diff
@@ -1,24 +1,70 @@
+
+| Column        | Type       | Options                        |
+| ------------- | ---------- | ------------------------------ |
+| card_info     | string     | null: false                    |
+| term          | string     | null: false                    |
+| security_code | string     | null: false                    |
```
```
| card_info     | string     | null: false                    |
| term          | string     | null: false                    |
| security_code | string     | null: false                    |
```

セキュリティの観点から、クレジットカードの情報はデータベースに直接保存しないようにしましょう。
つまり47~49行目は削除していただいて大丈夫です。

理由：
今回のフリマアプリではカードを登録する際は購入の度に入力となり、
データベースにわざわざ記録しなくても良い情報となるためです。

yuri240 reviewed (commented):
ER図をご提出いただき、ありがとうございます！
確認させていただきました。

purchaseテーブルに関して
２点コメントさせていただきましたので、ご確認お願いいたします。

sho1999 committed:
データベース設計の修正

・カード情報は毎回入力してもらうためデータベースに保存しなくてもOK
・誰が買ったか→user_id、何を買ったか→item_id
をpurchaseテーブルに記述する

sho1999 commented:
修正いたしました。

aoitech commented on README.md:
```diff
@@ -1,24 +1,71 @@
+| Column          | Type   | Options     |
+| --------------- | ------ | ----------- |
+| nickname        | string | null: false |
+| email           | string | null: false |
```
```ruby
| email           | string | null: false |
```
こちらTrelloの実装条件に「メールアドレスは一意性であること」とありますので、一意性制約をつけておきましょう。

aoitech commented on README.md:
```diff
@@ -1,24 +1,71 @@
+| --------------- | ------ | ----------- |
+| nickname        | string | null: false |
+| email           | string | null: false |
+| password        | string | null: false |
```
```ruby
| password        | string | null: false |
```
こちらのカラムはdeviseのgemを使用しての実装になるため、deviseが予め用意している
encrypted_passwordというカラムを用意してあげましょう。

aoitech commented on README.md:
```diff
@@ -1,24 +1,71 @@
+| first_nama      | string | null: false |
+| last_name_kana  | string | null: false |
+| first_nama_kana | string | null: false |
+| birthday        | string | null: false |
```
```ruby
| birthday        | string | null: false |
```
誕生日に関してですが、データ型は「date型」に修正しましょう。
（date型は年、月、日を１つで指定して取得できるためです。）

aoitech commented on README.md:
```diff
@@ -1,24 +1,71 @@
+| Column         | Type       | Options                        |
+| -------------- | ---------- | ------------------------------ |
+| name           | string     | null: false                    |
+| price          | string     | null: false                    |
```
```ruby
| price          | string     | null: false                    |
```
金額は数値なので、integer型にしましょう。

aoitech commented on README.md:
```diff
@@ -1,24 +1,71 @@
+| -------------- | ---------- | ------------------------------ |
+| name           | string     | null: false                    |
+| price          | string     | null: false                    |
+| category       | string     | null: false                    |
```
```ruby
| category       | string     | null: false                    |
| products_state | string     | null: false                    |
| delivery_fee   | string     | null: false                    |
| sending_region | string     | null: false                    |
| shipping_date  | string     | null: false                    |
```
商品の状態, 配送料の負担, 発送元の地域,  発送までの日数, カテゴリーのカラムはactive_hashで実装の予定のため、カラム名は単数形＋「_id」にしましょう。型はinteger型にしましょう。

aoitech commented on README.md:
```diff
@@ -1,24 +1,71 @@
-* Database creation
 
-* Database initialization
+## items テーブル
```
```ruby
## items テーブル
```
itemsテーブルに商品説明を保存するカラムが無いようです。
追加しましょう。
商品説明は長文になることも考えられるため、型はtext型にしましょう。

aoitech commented on README.md:
```diff
@@ -1,24 +1,71 @@
+| delivery_fee   | string     | null: false                    |
+| sending_region | string     | null: false                    |
+| shipping_date  | string     | null: false                    |
+| user_id        | references | null: false, foreign_key: true |
```
```ruby
| user_id        | references | null: false, foreign_key: true |
```
refernces型の場合は下記のようにuserで問題ございません。
もしカラム名で_idを使用する予定であれば、integer型を使用した方が良いでしょう。
```ruby
| user        | references | null: false, foreign_key: true |
```

aoitech commented on README.md:
```diff
@@ -1,24 +1,71 @@
+
+| Column        | Type       | Options                        |
+| ------------- | ---------- | ------------------------------ |
+| item_id       | references | null: false, foreign_key: true |
```
```ruby
| item_id       | references | null: false, foreign_key: true |
| user_id       | references | null: false, foreign_key: true |
```
こちらもreferences型の場合は`_id`を削除していただいて問題ございません。

aoitech commented on README.md:
```diff
@@ -1,24 +1,71 @@
+
+### Association
+
+- belongs_to :user
```
購入情報は１つの住所を持っている設計にしたいので、has_oneでアソシエーションを住所テーブルと組みましょう。

aoitech commented on README.md:
```diff
@@ -1,24 +1,71 @@
+- belongs_to :item
+
+
+## receivingテーブル
```
```ruby
## receivingテーブル
```
まだ必要なカラムが足りていないようでございます。
商品購入時に配送先住所を入力する際に必要な市町村、番地、建物番号、郵便番号、電話番号のカラムが用意されているかもう一度確認しましょう。

また、purchasesテーブルの外部キーカラムも用意しましょう。
ご不明点がございましたら、対応受付時間内にコネクトでご質問いただけますと幸いです。
お手数をお掛けいたしますが、よろしくお願いいたします。

aoitech replied:
```ruby
## receivingテーブル
```
こちらもテーブル名を複数形にしましょう。

aoitech commented on README.md:
```diff
@@ -1,24 +1,71 @@
+
+### Association
+
+- belongs_to :user
```
１つの住所は購入情報のテーブルに紐づく設計にしたいので、belongs_toでアソシエーションを購入情報のテーブルと組みましょう。
usersテーブルと住所テーブルとのアソシエーションは不要でございます。

aoitech commented on README.md:
```diff
@@ -1,24 +1,71 @@
+| Column         | Type       | Options     |
+| -------------- | ---------- | ----------- |
+| postcode       | string     | null: false |
+| receiving_area | string     | null: false |
```
```ruby
| receiving_area | string     | null: false |
```
今回都道府県のカラムはactive_hashを使用します。
active_hashのデータは使い回すことができますので、itemsテーブルの都道府県カラムと同じものを使用しましょう。型はinteger型になります。

aoitech commented on README.md:
```diff
@@ -1,24 +1,71 @@
+- belongs_to :purchase
+
+
+## purchaseテーブル
```
```ruby
## purchaseテーブル
```
細かい部分ではございますが、命名規則に則って複数形のpurchasesテーブルにしましょう。

sho1999 replied:
# なぜ
なぜ複数形になるのでしょうか？

aoitech commented on README.md:
```diff
@@ -1,24 +1,71 @@
 
-* Configuration
+- has_many :items
+- belongs_to :receiving
```
```ruby
belongs_to :receiving
```
今回住所テーブルとのアソシエーションを組まなくても、実装要件を満たすことができます。
削除しておきましょう。

aoitech commented on README.md:
```diff
@@ -1,24 +1,71 @@
-* Configuration
+- has_many :items
+- belongs_to :receiving
+- belongs_to :purchase
```
```ruby
belongs_to :purchase
```
userは何度でも商品を購入することが可能です。
そちらを踏まえてもう一度purchasesテーブルとの関係性を考えてみましょう。
ご不明点がございましたら、対応受付時間内にコネクトでご相談をお願いできますと幸いです。

aoitech commented on README.md:
```diff
@@ -1,24 +1,71 @@
 
-* ...
+- belongs_to :user
+- belongs_to :purchase
```
```ruby
belongs_to :purchase
```
今回1つの商品は1つの購入履歴を持っている設計にしたいため、上記のアソシエーションをもう一度見直してみましょう。

aoitech reviewed (commented):
レビュー依頼ありがとうございます。
コメントいたしましたので、ご確認いただけますと幸いです。

修正後、最新のER図も併せてご提出いただけますと幸いです。
その後再度レビュー依頼フォームのご提出をお願いいたします。
お手数をお掛け致しますが、よろしくお願いいたします。

sho1999 committed:
データベース修正2

・誕生日をdate型に
→年、月、日を１つで指定できるため
・商品の状態, 配送料の負担, 発送元の地域, 発送までの日数, カテゴリーのカラムはactive_hasuを使うためデータ型はintegarにする

sho1999 commented:
![スクリーンショット 2021-04-10 16 27 58](https://user-images.githubusercontent.com/63441286/114262230-b9fa9880-9a19-11eb-94a6-f6eb2c106b00.jpg)

yuri240 commented on README.md:
```diff
@@ -1,24 +1,79 @@
+| ------------------ | ------ | ----------- |
+| nickname           | string | null: false |
+| email              | string | null: false, unique: true |
+| password           | string | null: false |
```
```
| password           | string | null: false |
```
passwordカラムは不要なので、削除しましょう！

理由：
deviseのgemがあらかじめ用意している`encrypted_password`を使用しているため。

yuri240 commented on README.md:
```diff
@@ -1,24 +1,79 @@
+| block             | string     | null: false |
+| building          | string     | null: false |
+| phone_num         | string     | null: false |
+| item              | references | null: false, foreign_key: true |
+| user              | references | null: false, foreign_key: true |
```
71,72行目のカラムの代わりに
purchaseカラムを用意しましょう！

理由：
配送先住所が登録されるのと同時に
購入が行われるため。

yuri240 commented on README.md:
```diff
@@ -1,24 +1,79 @@
+
+### Association
+
+- belongs_to :user
+- belongs_to :item
```
77,78行目のアソシエーションは不要となりますので
削除しておきましょう！

yuri240 reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。

追加で３点コメントさせていただきましたので、ご確認お願いします。

sho1999 committed:
データベース修正2

aoitech commented on README.md:
```diff
@@ -1,24 +1,76 @@
 
-* System dependencies
+- has_many :items
+- belongs_to :receive
```
```ruby
belongs_to :receive
```
今回、receivesテーブルとのアソシエーションは記述しなくても実装要件を満たすことが可能です。
こちらは削除しておきましょう。

aoitech commented on README.md:
```diff
@@ -1,24 +1,76 @@
+| region_id         | integer    | null: false |
+| city              | string     | null: false |
+| block             | string     | null: false |
+| building          | string     | null: false |
```
```ruby
| building          | string     | null: false |
```
今回の実装では、建物名の入力は任意になるため、not null制約は削除しましょう。

aoitech reviewed (commented):
レビュー依頼ありがとうございます。
コメントいたしましたので、ご確認いただけますと幸いです。

修正後、最新のER図をご提出いただけますと幸いです。
その後再度レビュー依頼フォームのご提出をお願いいたします。

お手数をお掛け致しますが、よろしくお願いいたします。

sho1999 committed:
データベース修正3

sho1999 commented:
![Uploading スクリーンショット 2021-04-10 19.33.39.jpg…]()

sho1999 commented:
![スクリーンショット 2021-04-10 19 34 14](https://user-images.githubusercontent.com/63441286/114266879-bf64dc80-9a33-11eb-905a-be68076fc194.jpg)

Saka-toku commented:
レビュー依頼ありがとうございます！
修正点を確認いたしました。
ER図で以下の修正をお願いいたします。
```
・purchaseテーブルをpurchasesテーブルに名前を修正
・usersテーブルとpurchasesテーブルのアソシエーションの記述
・usersテーブルにencrypted_passwordカラムを追加
```
修正後は添付していただき、再度レビュー依頼フォームからご提出ください。

sho1999 committed:
ER図修正

sho1999 commented:
![スクリーンショット 2021-04-10 21 11 01](https://user-images.githubusercontent.com/63441286/114269300-466c8180-9a41-11eb-8f03-3a55dcdd4e3a.jpg)

kuu0123 commented:
レビュー依頼ありがとうございます！

ER図拝見いたしました。
1点気になる箇所がありますので、以下に記載いたします。

```
itemsテーブル
・商品説明に関するカラムが欠落してしまっているようです。追加しましょう。
```

修正が完了しましたら、フォームより改めてのレビュー依頼をお待ちしております！

sho1999 committed:
ER図修正

sho1999 commented:
![スクリーンショット 2021-04-11 14 52 13](https://user-images.githubusercontent.com/63441286/114293951-a14fb880-9ad5-11eb-94f9-fbdb6bbc98cc.jpg)

kata-myu reviewed (commented):
レビュー依頼ありがとうございます。
LGTMとさせていただきます。実装お疲れ様でした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き実装頑張っていきましょう。

sho1999 commented:
[テストコードと実行の動作.zip](https://github.com/sho1999/furima-32340/files/6291741/default.zip)
